content/blog/2022/2022-09-25-node-urlShortener.md

```diff
@@ -0,0 +1,494 @@
+- Click the **Confirm**  button.
+- Next, click on **Database** on the sidebar
+- Click the  **Connect**  button for your cluster
+- In the popup modal, click on  **Connect your application**.
+- **Copy** the URI on the clipboard
+- Lastly, all you need to do is replace the `<password>` field with the password you created previously.
+
+Our MongoDB atlas is all set and ready for use. Otherwise, click [here](https://www.freecodecamp.org/news/get-started-with-mongodb-atlas/) for a more in-depth guide on how to set up a MongoDB cluster.
+
+We can now, store our `DOMAIN_URL` and `MongoDB_URI` in `.env`:
+```
+MONGO_URI= mongodb+srv://<username>:<password>@cluster0.oq1hdin.mongodb.net/?retryWrites=true&w=majority
+DOMAIN_URL=http://localhost:3333
+```
+
+Enter the following code into the terminal to Start our Node.js server:
+```
+node index.js
+```
+Output:
+```
+Server is running at PORT: 3333
+```
+## Creating MongoDB Schema
+Everything in Mongoose starts with a Schema. Each schema maps to a MongoDB collection and defines the shape of the documents within that collection. We will need to create a Schema for our Urls. 
+
+Create a `Url.js` file in the `urlbackend` folder.
+
+To create our schema, paste the following code into `Url.js` file:
+```javascript
+const mongoose = require("mongoose");
+
+const UrlSchema = new mongoose.Schema({
+  urlId: {
+    type: String,
+    required: true,
+  },
+  origUrl: {
+    type: String,
+    required: true,
+  },
+  shortUrl: {
+    type: String,
+    required: true,
+  },
+  clicks: {
+    type: Number,
+    required: true,
+    default: 0,
+  },
+  date: {
+    type: String,
+    default: Date.now,
+  },
+});
+
+module.exports = mongoose.model("Url", UrlSchema);
+```
+In the above code we use mongoose to create a schema this will structure how the `Urls` are saved in our MongoDB database. To use schema definition, we converted our UrlSchema into a Model. By passing it into 
+
+`mongoose.model(modelName, schema)`
+
+## Create Helper Function To Validate Url Links
+
+Following that, we'll write a simple helper function. This will be used to validate that the URLs entered are in the right format. Create an `Util` folder and an `util.js` file in it.
+
+In the `util/util.js` file, add the following code:
+
+```javascript
+function validateUrl(value) {
+  var urlPattern = new RegExp('^(https?:\\/\\/)?'+ // validate protocol
+	    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // validate domain name
+	    '((\\d{1,3}\\.){3}\\d{1,3}))'+ // validate OR ip (v4) address
+	    '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*'+ // validate port and path
+	    '(\\?[;&a-z\\d%_.~+=-]*)?'+ // validate query string
+	    '(\\#[-a-z\\d_]*)?$','i');
+
+      return !!urlPattern.test(value);
+}
+
+module.exports = { validateUrl };
+``` 
+## Connecting to Database and Creating Endpoints
+In this section, we will connect the node js application to our database. And add all of the necessary endpoints to our application's `index.js` file.
```

Comment:
Before pasting the code, can you explain what we're trying to achieve? I.e. we want to have a POST endpoint `/short` that does this and a GET endpoint `/all` that does that ... .